### DISCUSSION TITLE
Patrol CLI commands are not recognized in Windows 11 and VSCode

### DISCUSSION THREAD

**Discussion by lmlikota** [leancodepl/patrol#412] [Q&A] [answered]
Hi guys,

really nice work with Patrol and thank you for the open source.  

I'm lucky to start with integration tests with Patrol and ditch integration_test package on my first script, I feel blessed already to get rid of `pumpAndSettle` which was pumping my nerves to much.

## The problem

I'm stuck, what I'm trying is to use `nativeAutomation: true` to tap on the permission on one of my onboarding screens.
![image](https://user-images.githubusercontent.com/10289319/194053639-7e093ae8-eb33-4579-b82d-8fc060a6e932.png)

I was running my tests using `flutter test` but if I understood it correctly when using `nativeAutomation: true`  I have to use `patrol drive`?

If that's correct I have an issue where Windows 11 don't recognize `patrol` command even though CLI is installed and I did all manually what `patrol bootstrap` would do.

### This is my first test
```
import 'package:flutter/cupertino.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:haven_bridge/main_common.dart';
import 'package:integration_test/integration_test.dart';
import 'package:patrol/patrol.dart';

import 'package:haven_bridge/main_qa.dart' as app;

void main() {
  patrolTest(nativeAutomation: true, 'sign in', (PatrolTester $) async {
    app.main();

    //expect($('Saznajte više'), findsOneWidget);
    await $('Prijava').tap();

    await $(#emailTextField).enterText(username);
    await $(#passwordTextField).enterText(password);
    await $(#loginButton).tap();
    expect($('Pregled intervencija'), findsOneWidget);
    await $.tap($('Sljedeće'));
    expect($('Odabir adrese'), findsOneWidget);
    await $.tap($('Sljedeće'));
    expect($('Voditelj intervencije'), findsOneWidget);
    await $.tap($('Sljedeće'));
    expect($('Uklanjanje vozila'), findsOneWidget);
    await $.tap($('Sljedeće'));
    expect($('Osvrt'), findsOneWidget);
    await $.tap($('Gotovo'));
    await $.pump(const Duration(seconds: 4));
    await $.native.grantPermissionWhenInUse();
  });
}
```
This is my console output from Powershell  when using `patrol drive` or any other `patrol` command

```
PS C:\Users\lmlikota\haven-bridge-mobile> patrol
patrol: The term 'patrol' is not recognized as a name of a cmdlet, function, script file, or executable program.
Check the spelling of the name, or if a path was included, verify that the path is correct and try again.

PS C:\Users\lmlikota\haven-bridge-mobile> patrol drive --flavor development .\integration_test\patrollogin.dart  
patrol: The term 'patrol' is not recognized as a name of a cmdlet, function, script file, or executable program.
Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
```
When I use `flutter test --flavor development .\integration_test\login.dart` in combination with `nativeAutomation: true`  I got this error in Powershell

```
PS C:\Users\lmlikota\haven-bridge-mobile> flutter test --flavor development .\integration_test\patrollogin.dart
  package:flutter_test/src/binding.dart                                         new _TestWidgetsFlutterBinding&BindingBase&SchedulerBinding  package:flutter_test/src/binding.dart                                         new _TestWidgetsFlutterBinding&BindingBase&SchedulerBinding&ServicesBinding
  package:flutter_test/src/binding.dart                                         new _TestWidgetsFlutterBinding&BindingBase&SchedulerBinding&ServicesBinding&GestureBinding
  package:flutter_test/src/binding.dart                                         new _TestWidgetsFlutterBinding&BindingBase&SchedulerBinding&ServicesBinding&GestureBinding&SemanticsBinding  package:flutter_test/src/binding.dart                                         new _TestWidgetsFlutterBinding&BindingBase&SchedulerBinding&ServicesBinding&GestureBinding&SemanticsBinding&RendererBinding
  package:flutter_test/src/binding.dart                                         new _TestWidgetsFlutterBinding&BindingBase&SchedulerBinding&ServicesBinding&GestureBinding&SemanticsBinding&RendererBinding&PaintingBinding
  package:flutter_test/src/binding.dart                                         new _TestWidgetsFlutterBinding&BindingBase&SchedulerBinding&ServicesBinding&GestureBinding&SemanticsBinding&RendererBinding&PaintingBinding&WidgetsBinding
  package:flutter_test/src/binding.dart                                         new _TestWidgetsFlutterBinding&BindingBase&SchedulerBinding&ServicesBinding&GestureBinding&SemanticsBinding&RendererBinding&PaintingBinding&WidgetsBinding&TestDefaultBinaryMessengerBinding
  package:flutter_test/src/binding.dart                                         new TestWidgetsFlutterBinding
  package:flutter_test/src/binding.dart                                         new LiveTestWidgetsFlutterBinding
  package:integration_test/integration_test.dart                                new IntegrationTestWidgetsFlutterBinding
  package:patrol/src/native/binding.dart 9:21                                   new PatrolBinding
  package:patrol/src/native/binding.dart 15:7                                   new PatrolBinding.ensureInitialized
  package:patrol/src/native/native_automator.dart 67:19                         new NativeAutomator.forTest
  package:patrol/src/custom_finders/common.dart 39:25                           patrolTest
  c:/Users/lmlikota/haven-bridge-mobile/integration_test/patrollogin.dart 10:3  main
```
When I remove `nativeAutomation: true` , comment out ` await $.native.grantPermissionWhenInUse();` and run test again using ` flutter test --flavor development .\integration_test\patrollogin.dart` test is working but off course **I'm stuck on permission screen**.

```
PS C:\Users\lmlikota\haven-bridge-mobile> flutter test --flavor development .\integration_test\patrollogin.dart
00:00 +0: loading C:\Users\lmlikota\haven-bridge-mobile\integration_test\patrollogin.dart                                                                                                                                              R00:36 +0: loading C:\Users\lmlikota\haven-bridge-mobile\integration_test\patrollogin.dart                                                                                                                                         36,5s
√  Built build\app\outputs\flutter-apk\app-development-debug.apk.
00:38 +0: loading C:\Users\lmlikota\haven-bridge-mobile\integration_test\patrollogin.dart                                                                                                                                              I00:46 +0: loading C:\Users\lmlikota\haven-bridge-mobile\integration_test\patrollogin.dart                                                                                                                                          7,6s
00:48 +0: sign in
[🌎 Easy Localization] [DEBUG] Localization initialized
[🌎 Easy Localization] [DEBUG] Start
[🌎 Easy Localization] [DEBUG] Init state
[🌎 Easy Localization] [INFO] Start locale loaded hr
[🌎 Easy Localization] [DEBUG] Build
[🌎 Easy Localization] [DEBUG] Init Localization Delegate
[🌎 Easy Localization] [DEBUG] Init provider
[🌎 Easy Localization] [DEBUG] Load Localization Delegate
Can't find an app in the Play Store with the id: hr.biss.haven_bridge.dev
01:02 +1: All tests passed!
```

Any help would be appreciated.

**Comment by bartekpacia** [comment, marked as answer]
Hi @lmlikota!

> I was running my tests using flutter test but if I understood it correctly when using nativeAutomation: true I have to use patrol drive?

That's right, you have to use `patrol drive`. It's required only if you use `$.native` (and I see you do). `patrol drive` does all the necessary setup work required to make tapping on native views, handling permission etc. work correctly.

If you only use our custom finders (the dollar – `$`), you should only `flutter test` :)

Now onto your problem. If after installing `patrol_cli` from pub.dev like this:

```
dart pub global activate patrol_cli
```

your system can't find the `patrol` executable, you most probably don't have pub cache added to your $PATH. Please refer to [these instructions](https://pub.dev/packages/patrol_cli#installation). Let me know if this solves your problem.


I'm so happy that you enjoy Patrol 💙 Thanks a lot you for your feedback. If you encounter any more problems are have questions, don't hesitate to ask :)

**Reply by lmlikota** [reply]
Thank you so much, I missed that part somehow 🙏. It's clearly stated in instructions 🤦‍♂️  Thank you for your kind and fast support, I will most definitely get stuck somewhere soon since I'm new to Dart and Flutter and kind of jumped on that train few days ago.

**Comment by bartekpacia** [comment]
Also, as a side note, I see that you're not passing anything to the `config` argument to `patrolTest()`. I suggest you use it :) For example, it'll make `$.native.openApp()` open your app if you don't pass anything as an argument. [See the example app's config here](https://github.com/leancodepl/patrol/blob/master/packages/patrol/example/integration_test/config.dart).

**Reply by lmlikota** [reply]
Thanks again 😄 I will definitely.

**Comment by lmlikota** [comment]
Hi @bartekpacia I'm stuck, again 🥲 

I have created `config` and I'm passing it to the `patrolTest()`

```dart
void main() {
  patrolTest(config: patrolConfig, nativeAutomation: true, 'sign in',
      (PatrolTester $) async {
```
I'm also using now ` await $.pumpWidgetAndSettle(MyApp());` instead of `app.main()`

What I'm getting now is really weird behavior. To explain what's going on I had to record it

https://user-images.githubusercontent.com/10289319/194548627-5a51c102-fbd3-4414-81d0-460e692c9790.mp4

and little bit different try

https://user-images.githubusercontent.com/10289319/194549591-1774325c-7e2b-41e5-99fc-1cd6c5dffebc.mp4

This is my `config.dart`
```dart
import 'package:patrol/patrol.dart';

// TODO: Replace with values specific to your app.
const patrolConfig = PatrolTestConfig(
  appName: '[QA] HAVEN FR',
  packageName: 'hr.biss.haven_bridge.dev',
  bundleId: 'hr.biss.havenBridge',
);
```
This is my console output from the video
PS C:\Users\lmlikota\haven-bridge-mobile> patrol drive --flavor development .\integration_test\patrollogin.dart

Update available! 0.6.11 → 0.6.12
Run patrol update to update

No device specified, using the first one (emulator-5554)
✓ Forwarded ports (0.2s)
✓ Installed server (0.9s)
✓ Installed instrumentation (1.7s)
Running app_test.dart with flutter_driver on emulator-5554...
pl.leancode.automatorserver.ServerLoop:
VMServiceFlutterDriver: Connecting to Flutter application at http://127.0.0.1:50164/6rC7218x7po=/
VMServiceFlutterDriver: Isolate found with number: 3048629238825619
VMServiceFlutterDriver: Isolate is paused at start.
VMServiceFlutterDriver: Attempting to resume isolate
Patrol: creating NativeAutomator, host: localhost, port: 8081, packageName: hr.biss.haven_bridge.dev, bundleId: hr.biss.havenBridge
00:00 +0: counter state is the same after going to home and switching apps
VMServiceFlutterDriver: Connected to Flutter application.
Patrol: pressHome() started
Patrol: pressHome() succeeded
Patrol: pressDoubleRecentApps() started
Patrol: pressDoubleRecentApps() succeeded
Patrol: openNotifications() started
Patrol: openNotifications() succeeded
Patrol: enableWifi() started
Patrol: enableWifi() succeeded
Patrol: disableWifi() started
Patrol: disableWifi() succeeded
Patrol: enableWifi() started
Patrol: enableWifi() succeeded
Patrol: pressBack() started
Patrol: pressBack() succeeded
00:15 +1: (tearDownAll)
00:15 +2: All tests passed!
flutter_driver exited with code 0
PS C:\Users\lmlikota\haven-bridge-mobile> flutter test --flavor development .\integration_test\patrollogin.dart
00:00 +0: loading C:\Users\lmlikota\haven-bridge-mobile\integration_test\patrollogin.dart                                                                       R00:40 +0: loading C:\Users\lmlikota\haven-bridge-mobile\integration_test\patrollogin.dart                                                                  40,4s
√  Built build\app\outputs\flutter-apk\app-development-debug.apk.
00:42 +0: loading C:\Users\lmlikota\haven-bridge-mobile\integration_test\patrollogin.dart                                                                       I00:48 +0: loading C:\Users\lmlikota\haven-bridge-mobile\integration_test\patrollogin.dart                                                                   6,3s
01:04 +0: sign in
══╡ EXCEPTION CAUGHT BY WIDGETS LIBRARY ╞═══════════════════════════════════════════════════════════
The following _CastError was thrown building MyApp(dirty):
Null check operator used on a null value

The relevant error-causing widget was:
  MyApp MyApp:file:///C:/Users/lmlikota/haven-bridge-mobile/integration_test/patrollogin.dart:14:33

When the exception was thrown, this was the stack:
#0      BuildContextEasyLocalizationExtension.localizationDelegates (package:easy_localization/src/public_ext.dart:148:32)
#1      MyApp.build (package:haven_bridge/main_common.dart:76:41)
#2      StatelessElement.build (package:flutter/src/widgets/framework.dart:4949:49)
#3      ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:4878:15)
#4      Element.rebuild (package:flutter/src/widgets/framework.dart:4604:5)
#5      ComponentElement._firstBuild (package:flutter/src/widgets/framework.dart:4859:5)
#6      ComponentElement.mount (package:flutter/src/widgets/framework.dart:4853:5)
#7      Element.inflateWidget (package:flutter/src/widgets/framework.dart:3863:16)
#8      Element.updateChild (package:flutter/src/widgets/framework.dart:3586:20)
#9      SingleChildRenderObjectElement.update (package:flutter/src/widgets/framework.dart:6307:14)
#10     Element.updateChild (package:flutter/src/widgets/framework.dart:3570:15)
#11     RenderObjectToWidgetElement._rebuild (package:flutter/src/widgets/binding.dart:1195:16)
#12     RenderObjectToWidgetElement.update (package:flutter/src/widgets/binding.dart:1172:5)
#13     RenderObjectToWidgetElement.performRebuild (package:flutter/src/widgets/binding.dart:1186:7)
#14     Element.rebuild (package:flutter/src/widgets/framework.dart:4604:5)
#15     BuildOwner.buildScope (package:flutter/src/widgets/framework.dart:2667:19)
#16     WidgetsBinding.drawFrame (package:flutter/src/widgets/binding.dart:882:21)
#17     RendererBinding._handlePersistentFrameCallback (package:flutter/src/rendering/binding.dart:378:5)
#18     SchedulerBinding._invokeFrameCallback (package:flutter/src/scheduler/binding.dart:1175:15)
#19     SchedulerBinding.handleDrawFrame (package:flutter/src/scheduler/binding.dart:1104:9)
#20     LiveTestWidgetsFlutterBinding.handleDrawFrame (package:flutter_test/src/binding.dart:1559:13)
#21     SchedulerBinding._handleDrawFrame (package:flutter/src/scheduler/binding.dart:1015:5)
#25     _invoke (dart:ui/hooks.dart:150:10)
#26     PlatformDispatcher._drawFrame (dart:ui/platform_dispatcher.dart:318:5)
#27     _drawFrame (dart:ui/hooks.dart:115:31)
(elided 3 frames from dart:async)

════════════════════════════════════════════════════════════════════════════════════════════════════
══╡ EXCEPTION CAUGHT BY FLUTTER TEST FRAMEWORK ╞════════════════════════════════════════════════════
The following WaitUntilVisibleTimeoutException was thrown running a test:
TimeoutException after 0:00:10.000000: Finder zero widgets with text "Prijava" (ignoring offstage
widgets) did not find any visible widgets

When the exception was thrown, this was the stack:
#0      PatrolTester.waitUntilVisible.<anonymous closure> (package:patrol/src/custom_finders/patrol_tester.dart:279:11)
<asynchronous suspension>
<asynchronous suspension>
(elided one frame from package:stack_trace)

The test description was:
  sign in
════════════════════════════════════════════════════════════════════════════════════════════════════
══╡ EXCEPTION CAUGHT BY FLUTTER TEST FRAMEWORK ╞════════════════════════════════════════════════════
The following message was thrown:
Multiple exceptions (2) were detected during the running of the current test, and at least one was
unexpected.
════════════════════════════════════════════════════════════════════════════════════════════════════
01:04 +0 -1: sign in [E]
  Test failed. See exception logs above.
  The test description was: sign in


To run this test again: C:\src\bin\cache\dart-sdk\bin\dart.exe test C:\Users\lmlikota\haven-bridge-mobile\integration_test\patrollogin.dart -p vm --plain-name 'sign in'
01:05 +0 -1: Some tests failed.
PS C:\Users\lmlikota\haven-bridge-mobile> flutter test --flavor development .\integration_test\patrollogin.dart
00:00 +0: loading C:\Users\lmlikota\haven-bridge-mobile\integration_test\patrollogin.dart                                                                       R00:38 +0: loading C:\Users\lmlikota\haven-bridge-mobile\integration_test\patrollogin.dart                                                                  38,7s
√  Built build\app\outputs\flutter-apk\app-development-debug.apk.
00:40 +0: loading C:\Users\lmlikota\haven-bridge-mobile\integration_test\patrollogin.dart                                                                       I00:45 +0: loading C:\Users\lmlikota\haven-bridge-mobile\integration_test\patrollogin.dart                                                                   5,5s
10:15 +0: loading C:\Users\lmlikota\haven-bridge-mobile\integration_test\patrollogin.dart
```

**Comment by bartekpacia** [comment]
Hi @lmlikota 

To keep the repo organized, please create a new issue and attach in it:

- all logs from your terminal (for both runs)
- your `PatrolConfig`
- your app's `main()` method

Thanks! :)

**Reply by lmlikota** [reply]
Sorry didn't see your comment before editing. I will create another discussion.

**Comment by bartekpacia** [comment]
I see you edited your post and attached logs. I see that something's wrong with EasyLocalization – didn't you forget to initialize it in your test?

PS `easy_localization` is a really bad library. We were using it at LeanCode in the past and it wasn't a good experience. We've migrated to Flutter's [native i18n solution](https://docs.flutter.dev/development/accessibility-and-localization/internationalization) and it works like a charm, I recommend you to do so!

**Reply by lmlikota** [reply]
Thanks, actually I'm tester 😄 I will pass this info to my Flutter devs.

**Reply by bartekpacia** [reply]
Ah, sure :)

`easy_localization` was OK when Flutter didn't have good i18n support, but now there's no need for it.

**Comment by bartekpacia** [comment]
@lmlikota Please create an Issue, not a Discussion :) If you have a problem with something not working, issues are the way to go.

**Reply by lmlikota** [reply]
@bartekpacia Should I delete this discussion entirely?

**Reply by bartekpacia** [reply]
No, there's a lot of useful info in here :). And btw, you shouldn't have permission to do so – do you? 👀

**Reply by lmlikota** [reply]
You're right, I don't have permission to delete discussion 😄
